[reland][dtensor] move DTensor to public namespace

wanchaol opened the pull request:
reland of https://github.com/pytorch/pytorch/pull/133113

I have to create a new PR because the previous reverted PR could not either be rebased, or imported successfully :(

----

Moving DTensor to be in the public namespace, to formally add the documentation page that includes all the public APIs. This includes:

* many path renames and path import fixes
* a dedicated doc page without too much content yet (adding in the next PRs)
* To preserve the BC for users still using the torch.distributed._tensor, I added a shim script to redirect old path calls to the new module

The BC preserving is evidented by the fact that all DTensor tests are still working without changing the public imports. So it's safe to land the changes


cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

albanD commented on docs/source/conf.py:
```diff
@@ -2621,6 +2632,15 @@
     # torch.distributed.rpc.rref_proxy
     "RRefProxy",
+    # torch.distributed.tensor.api
+    "DTensor",
```
If you have to add things here, it means it's not showing on the website.
You should not add anything to this file ever in general as it means there is some public API that is not documented being added.

If you properly updated the .rst to show your docstring, nothing is needed here.

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -231,6 +232,9 @@
   ],
   "torch.distributed.rpc": [],
+  "torch.distributed.tensor": [
+    "DeviceMesh"
```
New entries here are not allowed either. Please make the corresponding class properly public.

wanchaol replied:
done :)

seemethere reviewed (commented):
LGTM from a PR sanity POV, will let someone else approve who can code review in more detail

wanchaol committed:
[reland][dtensor] move DTensor to public namespace (#134203)

Summary:
reland of https://github.com/pytorch/pytorch/pull/133113

I have to create a new PR because the previous reverted PR could not either be rebased, or imported successfully :(

----

Moving DTensor to be in the public namespace, to formally add the documentation page that includes all the public APIs. This includes:

* many path renames and path import fixes
* a dedicated doc page without too much content yet (adding in the next PRs)
* To preserve the BC for users still using the torch.distributed._tensor, I added a shim script to redirect old path calls to the new module

The BC preserving is evidented by the fact that all DTensor tests are still working without changing the public imports. So it's safe to land the changes

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec LucasLLC MeetVadakkanchery mhorowitz pradeepfn

Pull Request resolved: https://github.com/pytorch/pytorch/pull/134203

Reviewed By: tianyu-l

Differential Revision: D61643864

Pulled By: wanchaol
